WreckEngine/Text.cpp

```diff
@@ -1,198 +1,571 @@
+    }
+    else
+    {
+        f = make_shared<FontFace>(font);
+        if (!f->fontFace)
+            return nullptr;
+        fontFaces.insert({font, f});
+    }
+
+    f->setSize(height, width);
+    return f;
+}
+
+void Text::FontFace::setSize(const uint32_t _height, const uint32_t _width)
+{
+    if (height == _height && width == _width)
+        return;
+
+    FT_Set_Pixel_Sizes(fontFace, width = _width, height = _height);
+
+    // Get the size of a space
+    FT_Glyph glyph = nullptr;
+    if (!FT_Load_Char(fontFace, ' ', FT_LOAD_DEFAULT) &&
+        !FT_Get_Glyph(fontFace->glyph, &glyph))
+    {
+        spaceWidth = fontFace->glyph->metrics.horiAdvance * kerningScale;
+    }
+
+    // Get the line height
+    lineHeight = fontFace->size->metrics.height * kerningScale;
+
+    // TODO - We may not want to immediately load glyphs here, or we may want to specify an overload for a glyph range
+    loadGlyphs();
+}
+
+void Text::FontFace::loadGlyphs()
+{
+    // Loads all basic, printable ASCII characters
+    loadGlyphRange(0, 255);
+}
+
+void Text::FontFace::loadGlyphRange(uint32_t begin, uint32_t end)
+{
+    // TODO - (M)SDF conversion?
+    // TODO - Currently assumes the texture is empty. Expected functionality would be to have
+    //        multiple glyph ranges loaded at a time
+
+    constexpr int rectPaddingPerSide = 1;
+    constexpr int rectPadding = rectPaddingPerSide * 2;
+
+    // Load each of the glyphs in the range
+    std::vector<stbrp_rect> packRects;
+    std::unordered_map<uint32_t, shared<GlyphData>> loadedGlyphs;
+    for (uint32_t cp = begin; cp <= end; ++cp)
+    {
+        // Load code point
+        if (FT_Load_Char(fontFace, cp, FT_LOAD_FORCE_AUTOHINT))
+        {
+            std::cout << "Could not load code point " << cp << " (" << static_cast<char>(cp) << ')' << std::endl;
+            continue;
+        }
+
+        // Load glyph of code point
+        FT_Glyph glyph = nullptr;
+        if (FT_Get_Glyph(fontFace->glyph, &glyph))
+        {
+            std::cout << "Failed to get glyph for " << cp << " (" << static_cast<char>(cp) << ')' << std::endl;
+            continue;
+        }
+
+        // Render glyph to bitmap
+        if (FT_Glyph_To_Bitmap(&glyph, FT_RENDER_MODE_NORMAL, 0, 1))
+        {
+            std::cout << "Failed to render glyph for " << cp << " (" << static_cast<char>(cp) << ')' << std::endl;
+            continue;
+        }
+        auto& bitmap = reinterpret_cast<FT_BitmapGlyph>(glyph)->bitmap;
+        const auto bitmapWidth = bitmap.width;
+        const auto bitmapHeight = bitmap.rows;
+
+        // Create pack rect (inflate it a bit for padding)
+        stbrp_rect rect = {0};
+        rect.id = cp;
+        rect.w = static_cast<stbrp_coord>(bitmapWidth + rectPadding);
+        rect.h = static_cast<stbrp_coord>(bitmapHeight + rectPadding);
+        packRects.push_back(rect);
+
+        // Create glyph data
+        auto glyphData = std::make_shared<GlyphData>();
```

Comment:
`GlyphData` contains a vector, but I guess since `loadGlyphRange` is called extremely infrequently the (most likely minor) hiccup caused by copying should be fine